store: Add node Ephemeral support

abronan opened the pull request:
Adds an `Ephemeral` flag to Consul, Zookeeper and Etcd for usage with Discovery.

This PR is mostly made to discuss about what is the best approach: rely on TTLs or use an Ephemeral flag. In both cases there are trade-offs to be made on the semantics of both operations:

- If we use an Ephemeral flag, we hide etcd's lack of support for Ephemeral by the use of a periodic `Set` that renews the TTL (What this PR does)
- If we use a TTL, we must hide zookeeper's lack of support for TTL behind the use of an Ephemeral flag.

abronan commented:
partially fixes #794

aluzzardi commented on cli/join.go:
```diff
@@ -42,11 +42,10 @@ func join(c *cli.Context) {
+	if err := d.Register(addr); err != nil {
+		log.Error(err)
+	}
 	for {
```
`select{}`

aluzzardi replied:
Actually you were right. I would keep this the way it was and continue issuing periodic `Register()` in case something goes wrong (instead of doing just one Register)

aluzzardi commented on pkg/store/store.go:
```diff
@@ -97,6 +97,11 @@ type KVPair struct {
 }
 
+// RequestOptions contains optional request parameters
+type RequestOptions struct {
```
`WriteOptions`?

aluzzardi replied:
`AtomicPut` should take a `WriteOptions` as well, to be consistent.

aluzzardi replied:
Also, ephemeral should probably be implemented in AtomicPut as well I guess

aluzzardi commented on cli/join.go:
```diff
@@ -42,11 +42,10 @@ func join(c *cli.Context) {
 	}
 
 	hbval := time.Duration(hb)
+	log.WithFields(log.Fields{"addr": addr, "discovery": dflag}).Infof("The entry in the discovery will disappear if the node goes down..", hbval)
```
Probably not worth mentioning

abronan replied:
Ok

aluzzardi commented on pkg/store/etcd.go:
```diff
@@ -99,13 +86,21 @@ func (s *Etcd) Get(key string) (*KVPair, error) {
+func (s *Etcd) Put(key string, value []byte, opts *RequestOptions) error {
+
+	// Default TTL = 0 means no expiration
+	ttl := 0 * time.Second
```
`ttl := 0` should be enough?

aluzzardi commented on pkg/store/etcd.go:
```diff
@@ -99,13 +86,21 @@ func (s *Etcd) Get(key string) (*KVPair, error) {
+	// Default TTL = 0 means no expiration
+	ttl := 0 * time.Second
+	if opts != nil && opts.Ephemeral {
+		ttl = 60 * time.Second
```
Make `60` either:

- a `const`
- a `WriteOption`
- a `Config`

abronan replied:
If I put this into `WriteOption` that would defeat the purpose of hiding it behind the `Ephemeral` flag. Same thing if I put this in Config. Best bet for now would be to put those as `const` (unless we can specify `heartbeat` + `refresh` time in the `cli`).

aluzzardi commented on pkg/store/etcd.go:
```diff
@@ -115,6 +110,30 @@ func (s *Etcd) Put(key string, value []byte) error {
+
+// Periodically renew a key with an Ephemeral flag set
+func (s *Etcd) periodicRenew(key string, value []byte, ttl time.Duration) {
+	renew := time.NewTicker(40 * time.Second)
```
Same here:

Make `40` either:
- a `const`
- a `WriteOption`
- a `Config`

This could be named `heartbeat` or something like that (this is exactly what the heartbeat was supposed to be like)

aluzzardi replied:
Instead of `NewTicket`, you can use:

```
renewCh := time.Tick(...)
for _ = range renew {
}
```

aluzzardi commented on pkg/store/etcd.go:
```diff
@@ -115,6 +110,30 @@ func (s *Etcd) Put(key string, value []byte) error {
+func (s *Etcd) periodicRenew(key string, value []byte, ttl time.Duration) {
+	renew := time.NewTicker(40 * time.Second)
+	for _ = range renew.C {
+		if _, err := s.client.Set(key, string(value), uint64(ttl)); err != nil {
```
This will override the key, meaning that:

```
c.Put("foo", "bar", &WriteOptions{Ephemeral: true})
c.Put("foo", "hello world")
c.Get("foo") <- after 40 seconds this is "bar" again
```

You can solve this by using an `CompareAndSwap` and if it fails, at that point do a get. 
Probably no need to get the `value` as a parameter: just go a get first, grab the lastIndex and value, then compare and swap with that value and last index. When it fails, get the new value and lastIndex and loop on

abronan replied:
I think the good practice would be to use the `Update` method of the go-client, so I need to dig this.

abronan replied:
Well cannot seem to avoid the CAS call here: there is no way with the client to only update the key's TTL without specifying a value, this is only do-able for a directory. So.. going back to the CAS solution..

aluzzardi commented on pkg/store/etcd.go:
```diff
@@ -99,13 +86,21 @@ func (s *Etcd) Get(key string) (*KVPair, error) {
+	ttl := 0 * time.Second
+	if opts != nil && opts.Ephemeral {
+		ttl = 60 * time.Second
+		go s.periodicRenew(key, value, ttl)
```
Unfortunately, I think we have to keep track of our goroutines :/

This means that every time we do an ephemeral put of the same key, we end up with a new goroutine.

We could use a kind of "set" (`map[string]struct{}`) and conditionally spawn goroutines

abronan commented:
Rebased and updated Godeps (there was an issue with `go-etcd` and the `ttl` parameter of `Set`, setting it to a value different than 0 would cause an index out of bound exception).

It now works consistently across the 3 stores (at least locally, Jenkins fails on `etcd` for some reasons). I kept it simple, for now the TTL is set as a const: `60s`, and it is hidden behind the `ephemeral` behavior. I'm waiting for #805 to update this and place the `EphemeralTTL` to be higher than the heartbeat (something like 2 or 3 times the heartbeat set on `join`).

abronan commented:
Should we update the `etcd` image on integration tests to use at least `v2.0.0` (and require everyone to use etcd >= `v2.0.0`)? I think that updating Godeps with the latest version of `go-etcd` is probably the reason integration tests are failing now.. (we still use `v0.4.6`)

I did all my tests on the latest version of etcd: `v2.0.11`

/cc @aluzzardi

aluzzardi commented:
@abronan I actually tried that - for some reason, updating etcd or go-etcd breaks our tests :(

I didn't try to updating both though.

We could try once we get #805 in since that PR is changing the tests a lot (they test way more), as it would be unsafe to upgrade without getting those tests in first.

abronan commented:
@aluzzardi Yes that was what I was thinking. I'll try updating both once we have your PR merged and see how we can fix this.

aluzzardi commented:
We should test that the TTL is working correctly with a test similar to #811

aluzzardi commented:
>  I kept it simple, for now the TTL is set as a const: 60s, and it is hidden behind the ephemeral behavior. I'm waiting for #805 to update this and place the EphemeralTTL to be higher than the heartbeat (something like 2 or 3 times the heartbeat set on join).

In order to be able to do integration tests, we'll need to be able to configure the TTL.

How about adding `EphemeralTTL` and `EphemeralHeartbeat` to the store `Config`?

We'd use the cli `heartbeat` flag for the heartbeat, and we could probably add a `swarm.discovery.ttl` flag for the TTL - it makes sense for other stores as well (`token` /cc @vieux)

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

abronan commented:
@aluzzardi Rebased once again. Tests are now passing. I changed the etcd image for the latest version `v2.0.11`. Still a few changes to make but we are slowly getting there :ghost: 

Also I think I'd probably like to add the integration tests at this point.

abronan commented:
ping @aluzzardi 

Rebased on master (with Node removal). It works fine, I added a single integration test for each backend.

It is not ideal at the moment but we need to figure out what to do with etcd first if we want a refresh happening in the background (because of the lack of a `TTL Update` function on key without specifying a value).

Also Consul does not accepts TTLs inferior to `10s`, thus the total integration tests time is increasing a little bit :-(

aluzzardi commented on discovery/discovery.go:
```diff
@@ -85,7 +85,7 @@ func (e Entries) Diff(cmp Entries) (Entries, Entries) {
 type Discovery interface {
 	// Initialize the discovery with URIs and a heartbeat.
-	Initialize(string, time.Duration) error
+	Initialize(string, time.Duration, time.Duration) error
```
Please update the comment above

abronan replied:
+1

aluzzardi commented on cli/flags.go:
```diff
@@ -53,6 +53,11 @@ var (
 		Usage: "period between each heartbeat",
 	}
+	flTTL = cli.StringFlag{
+		Name:  "time-to-live, ttl",
```
Should we call this just `ttl`?

Also, what about removing `hb` - just `heartbeat`?

I think shortcut helpers don't make much sense for advanced options

aluzzardi replied:
Also, how about rounding out?

What do you think of those defaults:
- `heartbeat` = 20s
- `ttl` = 60s

abronan replied:
+1

aluzzardi commented on cluster/swarm/cluster.go:
```diff
@@ -68,7 +68,7 @@ func NewCluster(scheduler *scheduler.Scheduler, store *state.Store, TLSConfig *t
 
 	// Set up discovery.
-	cluster.discovery, err = discovery.New(dflag, heartbeat)
+	cluster.discovery, err = discovery.New(dflag, heartbeat, 0)
```
Shouldn't we pass a `ttl` here?

aluzzardi replied:
Oh, I see - it's useless.

aluzzardi replied:
How about building the discovery object in manage directly? Dependency injection would be better.

Also, `swarm.discovery.heartbeat`: we could simply reuse the flHeartbeat we already have in manage

aluzzardi commented on discovery/discovery.go:
```diff
@@ -133,12 +133,12 @@ func parse(rawurl string) (string, string) {
 // New returns a new Discovery given a URL and heartbeat settings.
 // Returns an error if the URL scheme is not supported.
-func New(rawurl string, heartbeat time.Duration) (Discovery, error) {
+func New(rawurl string, heartbeat time.Duration, ttl time.Duration) (Discovery, error) {
```
Comment above

aluzzardi commented on pkg/store/consul.go:
```diff
@@ -91,11 +105,55 @@ func (s *Consul) Get(key string) (*KVPair, error) {
+			// Lock and ignore if lock is held
+			// It's just a placeholder for the
+			// ephemeral behavior
+			lock, _ := s.client.LockOpts(lockOpts)
```
I don't understand this?

abronan replied:
We need a lock as the holder of the session with `Behavior: Delete` to mimic Zookeeper's ephemeral node behavior :)

aluzzardi commented on pkg/store/zookeeper.go:
```diff
@@ -22,10 +22,13 @@ type zookeeperLock struct {
 func InitializeZookeeper(endpoints []string, options *Config) (Store, error) {
 	s := &Zookeeper{}
-	s.timeout = 5 * time.Second // default timeout
+	s.timeout = 10 * time.Second // default timeout
```
please put a const in `zookeeper.go` (private, doesn't need to be exported)

aluzzardi commented on pkg/store/consul.go:
```diff
@@ -91,11 +105,55 @@ func (s *Consul) Get(key string) (*KVPair, error) {
+		Value: value,
+	}
+
+	if opts != nil && opts.Ephemeral {
```
Since the session shares the same `ttl` as everyone else, I think you can create just one single session for all ephemeral keys.

Also, I would put this code in something like `createEphemeralSession` so you can remove it from the put.

Maybe you can call that function if `opts.Epheneral` and let it check if it needs or not to create a session:
```
if opts != nil && opts.Ephemeral {
   s.createEphemeralSession()
   [...]
```

createEphemeralSession could use a mutex to avoid creating multiple sessions for nothing.

aluzzardi commented on pkg/store/consul.go:
```diff
@@ -91,11 +105,55 @@ func (s *Consul) Get(key string) (*KVPair, error) {
+		if _, ok := s.sessions[key]; !ok {
+			entry := &api.SessionEntry{
+				Behavior: api.SessionBehaviorDelete,
+				TTL:      s.ephemeralTTL.String(),
```
maybe: `"kv_ephemeral_" + s.ephemeralTTL.String()`? Just to avoid clashing with user sessions

abronan replied:
Session ID is a generated uuid and the name to give to the session is optional. So we can add a name but I'm not sure it will serve any purpose.

aluzzardi commented on test/integration/discovery/discovery_helpers.bash:
```diff
@@ -10,6 +10,11 @@ function discovery_check_swarm_info() {
 	docker_swarm info | grep -q "Nodes: $count"
 }
 
+# Returns true if swarm info outputs is empty (0 nodes).
```
Actually, I changed the above (`discovery_check_swarm_info`) to take a parameter so you could do this with:

`discovery_check_swarm_info "$DISCOVERY" 0`

abronan replied:
Oh yeah I see that's new :)

aluzzardi commented on test/integration/discovery/consul.bats:
```diff
@@ -59,6 +59,31 @@ function teardown() {
 	retry 5 1 discovery_check_swarm_info
 }
 
+@test "consul discovery: check for engines departure" {
```
Could the name be consistent with `file`?
https://github.com/abronan/swarm/blob/add_ephemeral_support/test/integration/discovery/file.bats#L56

Either call this "node removal" or change file to "check for engines departure"

aluzzardi commented on test/integration/discovery/etcd.bats:
```diff
@@ -60,6 +64,31 @@ function teardown() {
 	retry 5 1 discovery_check_swarm_info
 }
 
+@test "etcd discovery: check for engines departure" {
```
Same comments than consul

aluzzardi commented on test/integration/discovery/zk.bats:
```diff
@@ -59,6 +59,31 @@ function teardown() {
 	retry 10 1 discovery_check_swarm_info
 }
 
+@test "zk discovery: check for engines departure" {
```
Same

aluzzardi commented on test/integration/discovery/consul.bats:
```diff
@@ -59,6 +59,31 @@ function teardown() {
+	retry 5 1 discovery_check_swarm_info
+
+	# Removes all the swarm agents
+	swarm_join_cleanup
```
So, with `file` I removed only one engine.

It's up to you what you prefer to do (remove one or all).

If you wish to remove only one, you could do `kill ${SWARM_JOIN_PIDS[0]}`

aluzzardi commented:
Few minor comments otherwise looks good

abronan commented:
@aluzzardi PTAL

/cc @vieux

aluzzardi commented on pkg/store/consul.go:
```diff
@@ -91,11 +125,47 @@ func (s *Consul) Get(key string) (*KVPair, error) {
+
+		// Create lock option with the
+		// EphemeralSession
+		lockOpts := &api.LockOptions{
```
Should we do this in `createEphemeralSession`?

abronan replied:
After tinkering a little bit with Consul and the lock mechanism, if we give the same session to the `api.KVPair` and `api.LockOptions`, the call to `lock.Lock()` is non blocking because of those lines in `consul/lock.go`:

```
if pair != nil && pair.Session == l.lockSession {
		goto HELD
}
```
I don't think we can put it in `createEphemeralSession` either, we need to go through the lock call each time just in case the session expired and we want to come back to the discovery service if we recover (but I need to add one line during an error of refresh if the session does not exists anymore and put it back to `""`).

Also, I thought that this might leak goroutines by just calling `lock.Lock()` each time because of this line in `consul/lock.go`:

```
go l.monitorLock(l.lockSession, leaderCh)
```
But it appears that no :D

It's just a placeholder lock that does nothing else but hold the ephemeral session :)

abronan replied:
Errr didn't answered the real question :D I think it's good here because we associate the key in `LockOptions` with the same global session. `LockOptions` should be different for each key though (to be associated with this global session).

aluzzardi commented on pkg/store/consul.go:
```diff
@@ -91,11 +125,47 @@ func (s *Consul) Get(key string) (*KVPair, error) {
+	if opts != nil && opts.Ephemeral {
+		// Creates the global ephemeral session
+		// if it does not exist
+		if s.ephemeralSession == "" {
```
We should probably not do the check here since it's already done by `createEphemeralSession`

aluzzardi commented:
LGTM
